src/relax/ir/dataflow_matcher.cc

```diff
@@ -523,109 +523,117 @@ bool MatchExpr(DFPattern pattern, Expr expr, Optional<Map<Var, Expr>> bindings_o
+  void add(const PNode* p, const RNode* r) {
+    match_p_r[p] = r;
+    match_r_p[r] = p;
+  }
 
-  const auto try_match_update_undo = [&](PNode* p, RNode* r) {
-    if (auto undo_more = try_match(p, r, m, def2use, use2def)) {
-      undo.insert(undo.end(), undo_more->begin(), undo_more->end());
-      return true;
+  void add(const MatchState& other) {
+    for (const auto& [p, r] : other.match_p_r) {
+      add(p, r);
     }
-    return false;
-  };
+  }
 
-  commit(p, r);
+  const VarNode* matched(const PNode* p) const {
+    if (!match_p_r.count(p)) return nullptr;
+    return match_p_r.at(p)->ptr;
+  }
 
-  // match parent patterns.
-  for (auto& [pparent, constraints] : p->parents) {
-    bool any_cons_sat = false;
-    for (auto& rparent : r->parents) {
-      // skip if mismatch.
-      if (rparent->matched && rparent->matched != pparent->ptr) continue;
+  const DFPatternNode* matched(const RNode* r) const {
+    if (!match_r_p.count(r)) return nullptr;
+    return match_r_p.at(r)->ptr;
+  }
 
-      const auto& uses = def2use.at(rparent->ptr);
+  const VarNode* matched(const PNode& p) const { return matched(&p); }
+  const DFPatternNode* matched(const RNode& r) const { return matched(&r); }
 
-      // check edge constraints.
-      bool cons_sat = true;
-      for (const auto& cons : constraints) {
-        if (cons.type == PairCons::kOnlyUsedBy && uses.size() != 1) {
-          cons_sat = false;
-          break;
-        }
-
-        if (cons.index != -1) {
-          const auto& callees = use2def.at(r->ptr);
-          if (callees.size() <= static_cast<size_t>(cons.index) ||
-              callees[cons.index] != rparent->ptr) {
-            cons_sat = false;
-            break;
-          }
-        }
-      }
-      if (!cons_sat) continue;
-      any_cons_sat = true;
+ private:
+  std::unordered_map<const PNode*, const RNode*> match_p_r;
+  std::unordered_map<const RNode*, const PNode*> match_r_p;
+};
 
-      // try all parent R nodes that are not matched yet.
-      // as long as ppattern can match one node.
-      if (!pparent->matched && try_match_update_undo(pparent, rparent)) {
-        commit(pparent, rparent);
-        break;
+/**
+ * \brief This method try to match a real node and a pattern node along with its neighbors.
+ */
+static std::optional<MatchState> TryMatch(const PNode& p, const RNode& r, DFPatternMatcher* m,
+                                          const MatcherUseDefAnalysis& ud_analysis) {
+  if (!m->Match(GetRef<DFPattern>(p.ptr), GetRef<Var>(r.ptr))) return std::nullopt;
+
+  MatchState result;
+
+  for (size_t i = 0; i < p.parents.size(); ++i) {
+    const auto p_node_parent = p.parents[i].first;
+    if (p_node_parent->ptr->IsInstance<WildcardPatternNode>()) {
+      ICHECK_EQ(p.parents.size(), r.parents.size());
+      if (auto v = result.matched(p_node_parent); v && v != r.parents[i]->ptr) {
+        // A parent wildcard pattern is already matched to other variable.
+        return std::nullopt;
```

Comment:
Sorry if I missed anything here. But could you kindly help explain why we need to add special handlings for nodes whose parents (aka arguments) are wildcard?

Reply:
I guess you meant that if it is wildcard, we don't have to obey "one-one" mapping (originally suggested here https://github.com/tlc-pack/relax/issues/160#issue-1268184762 in the `One-One Match` section). If that's the case, I felt it might still make some sense to optionally allow one-one mapping even for wildcards?

Reply:
This is for preventing false matching below:

```
 lv0 = R.matmul(x1, w0)
 lv1 = R.matmul(x2, w1)
 lv2 = R.matmul(x2, w2)
```

```
inp_pat = wildcard()
Q_weight_pat = wildcard()
K_weight_pat = wildcard()
V_weight_pat = wildcard()

matmul1 = is_op("relax.matmul")(inp_pat, Q_weight_pat)
matmul2 = is_op("relax.matmul")(inp_pat, K_weight_pat)
matmul3 = is_op("relax.matmul")(inp_pat, V_weight_pat)
```

The pattern adds a constraint that the LHS of three matmuls be the same. After matching the first matmul, we learn that `inp_pat` matches with `x1`. On the second matmul pattern + the binding (`lv1`), `m->Match(...)` would return true , since it simply matches a matmul pattern with two wildcards. This additional check tells us that `inp_pat` has already been matched to `x1`, so the second matching where `inp_pat` would match to `x2` must be disallowed.

Reply:
> The pattern adds a constraint that the LHS of three matmuls be the same.

Correct as we assume one DFPattern should only match one actual/real node.

> On the second matmul pattern + the binding (lv1), m->Match(...) would return true , since it simply matches a matmul pattern with two wildcards.

In theory, we want it to return `false` due to one-one mapping. IIRC, I thought when matching the 2nd mm pattern to `R.matmul(x2, w1)`, we will check if `x2` is the the matched RNode for `inp_pat` (which is `x1`) => so false. That being said, I don't understand why we need to specifically check the `wildcard` pattern. Because in theory if a pattern is matched to an RNode, and when re-matching the same pattern we shall assert that the incoming RNode must be the matched one.

Reply:
> In theory, we want it to return false due to one-one mapping

Note that by `m->Match(...)` I meant a pure DF pattern matcher. It is not aware of RNode etc.

> That being said, I don't understand why we need to specifically check the wildcard pattern

Right, this additional condition is not specific to wildcard. I lifted this check out of `if (p_node_parent->ptr->IsInstance<WildcardPatternNode>()) {` and I'm trying to see if I can remove this if block (L619-L622) altogether.

Reply:
Good news, I found two bugs in my implementation. They are fixed in the latest commit, and this entire `if` block has been removed.